### PR TITLE
EWC-225 progress

### PR DESCRIPTION
**Pull request by Railag**
EWC-225 progress: transactions for user endpoints, updated baseManager, logout and activate user endpoints

### CODE REVIEW

File: src/controllers/user-controller.js [hunk 1]
```diff
@@ -39,8 +40,22 @@ const resendActivationEndPoint = async function (req) {
...
+  return await UserService.logout(user);
+};
+
 module.exports = {
   userSignupEndPoint: userSignupEndPoint,
   userLoginEndPoint: userLoginEndPoint,
-  resendActivationEndPoint: resendActivationEndPoint
+  resendActivationEndPoint: resendActivationEndPoint,
+  activateUserAccountEndPoint: activateUserAccountEndPoint,
+  userLogoutEndPoint: AuthDecorator.checkAuthToken(_userLogoutEndPoint)
```
Comment (baghbidi) [inline src/controllers/user-controller.js]: trailing comma

File: src/controllers/user-controller.js [hunk 2]
```diff
@@ -39,8 +40,22 @@ const resendActivationEndPoint = async function (req) {
   return await UserService.resendActivation(emailData);
 };
 
+const activateUserAccountEndPoint = async function (req) {
+  logger.info("Parameters:" + JSON.stringify(req.body));
```
Comment (baghbidi) [inline src/controllers/user-controller.js]: `req.Body` contains pii, so should not be logged without encryption/hashing.

File: src/services/user-service.js [hunk 3]
```diff
@@ -45,61 +43,96 @@ const signUp = async function (user) {
 
   if (emailActivation === 'on') {
 
-    const newUser = await _handleCreateUser(user, emailActivation);
+    const newUser = await _handleCreateUser(user, emailActivation, transaction);
```
Comment (baghbidi) [inline src/services/user-service.js]: Do you encrypt password?

File: src/services/user-service.js [hunk 4]
```diff
@@ -45,61 +43,96 @@ const signUp = async function (user) {
...
 
-    const activationCodeData = await EmailActivationCodeService.generateActivationCode();
-    await EmailActivationCodeService.saveActivationCode(newUser.id, activationCodeData);
+    const activationCodeData = await EmailActivationCodeService.generateActivationCode(transaction);
+    await EmailActivationCodeService.saveActivationCode(newUser.id, activationCodeData, transaction);
 
-    const emailData = await _getEmailData();
+    const emailData = await _getEmailData(transaction);
     const transporter = await _userEmailSender(emailData);
     await _notifyUserAboutActivationCode(user.email, "https://google.com", emailData, activationCodeData, transporter);
```
Comment (baghbidi) [inline src/services/user-service.js]: `google.com`?

File: src/sequelize/managers/base-manager.js [hunk 5]
```diff
@@ -11,67 +11,67 @@
...
  */
 
-const moment = require('moment');
-const _ = require('underscore');
+const AppHelper = require('../../helpers/app-helper');
 
 module.exports = class BaseManager {
 
   getEntity() {
     return null;
```
Comment (baghbidi) [inline src/sequelize/managers/base-manager.js]: It should throw a meaningful (not implemented, ...) error here.

File: src/routes/user.js [hunk 6]
```diff
@@ -95,10 +107,28 @@ module.exports = [
...
+        {
+          code: constants.HTTP_CODE_NOT_FOUND,
+          errors: [Errors.NotFoundError]
+        }
+      ];
+
+      const activateUserEndPoint = ResponseDecorator.handleErrors(UserController.activateUserAccountEndPoint, successCode, errorCodes);
+      const responseObject = await activateUserEndPoint(req);
+
+      // redirect to login page
```
[the remark this answers is not in the document]
Comment (Railag) [reply src/routes/user.js]: but we still need to redirect user somewhere after email activation (clickable link in email looks like GET request to our API with accessToken query). We can move it to config, e.g. "homeUrl". Currently "google.com" just replaced previous IOAuthoring url in all email templates & this redirect

[the remark this answers is not in the document]
Comment (Railag) [reply src/routes/user.js]: moved url to config for now

File: src/services/user-service.js (hunk 3 above)
Comment (Railag) [reply src/services/user-service.js]: added password encryption

**Pull request merged**